From bbbe93842df03a16a457656910ddd163bb38d43b Mon Sep 17 00:00:00 2001
From: Edward Wu <edwardwu@realtek.com>
Date: Fri, 8 Nov 2024 11:19:00 +0800
Subject: [PATCH] CHROMIUM: soc: realtek: common: rpc bind CMA memory-region to
 limit the range of address

jira: CHROME-113, SW-7823

Chromium/AIOT don't need Secure Media Path
No need to use RTK media/dma heap driver.

Calling dma_alloc_coherent() could use system default CMA directly.
The range of address could be limited by CMA reserved memory.

However, Linux upstream will use alloc_page instead of cma_alloc if your
requesting size is less than or equal 4K

dma_alloc_contiguous()
{
...
        if (dev->cma_area)
                return cma_alloc_aligned(dev->cma_area, size, gfp);
        if (size <= PAGE_SIZE)
                return NULL;
...
        return cma_alloc_aligned(dma_contiguous_default_area, size, gfp);
}

alloc_page can't limit address and causes unlimited address returned.
RTK a/v/hifi fw can use limit address range, it causes fw misbehavior then.

Use dts and of_reserved_mem_device_init() to binding CMA. Our device has cma_area member.
So could go into the first condition cma_alloc_aligned() even size is less than or equal 4K.

Change-Id: I1b294f57bd688783044e46a8d3f0191652d4af83
---
 arch/arm64/boot/dts/realtek/rtd16xxb.dtsi | 3 +++
 drivers/soc/realtek/common/rtk_rpc_mem.c  | 8 +++++++-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/realtek/rtd16xxb.dtsi b/arch/arm64/boot/dts/realtek/rtd16xxb.dtsi
index a84f8f1936266..5cf2d783e768e 100644
--- a/arch/arm64/boot/dts/realtek/rtd16xxb.dtsi
+++ b/arch/arm64/boot/dts/realtek/rtd16xxb.dtsi
@@ -1511,6 +1511,7 @@ acpu_kernel_agent: acpu-kernel-agent {
 		acpu-intr-channel{
 			compatible = "realtek,rpc-intr-acpu";
 			name = "acpu-intr-channel";
+			memory-region = <&cma_reserved_0>;
 			tx-info = <0x4100100>;
 			rx-info = <0x4100140>;
 			tx-fifo = <0x40ff200>;
@@ -1552,6 +1553,7 @@ vcpu_kernel_agent: vcpu-kernel-agent {
 		vcpu-intr-channel {
 			compatible = "realtek,rpc-intr-vcpu";
 			name = "vcpu-intr-channel";
+			memory-region = <&cma_reserved_0>;
 			tx-info = <0x4100180>;
 			rx-info = <0x41001c0>;
 			tx-fifo = <0x40ffa00>;
@@ -1583,6 +1585,7 @@ hifi_kernel_agent: hifi-kernel-agent {
 		hifi-intr-channel {
 			compatible = "realtek,rpc-intr-hifi";
 			name = "hifi-intr-channel";
+			memory-region = <&cma_reserved_0>;
 			tx-info = <0x4102400>;
 			rx-info = <0x4102580>;
 			tx-fifo = <0x4101a00>;
diff --git a/drivers/soc/realtek/common/rtk_rpc_mem.c b/drivers/soc/realtek/common/rtk_rpc_mem.c
index 34727c40d7e38..6be84ec845135 100644
--- a/drivers/soc/realtek/common/rtk_rpc_mem.c
+++ b/drivers/soc/realtek/common/rtk_rpc_mem.c
@@ -10,6 +10,7 @@
 #include <linux/module.h>
 #include <linux/platform_device.h>
 #include <linux/of_platform.h>
+#include <linux/of_reserved_mem.h>
 #include <linux/io.h>
 #include <linux/interrupt.h>
 #include <linux/slab.h>
@@ -774,7 +775,7 @@ static int rtk_rpc_callback(struct rpmsg_device *rpdev,
 static int rtk_rpc_probe(struct rpmsg_device *rpdev)
 {
 	struct rtk_rpc_client *client;
-	int *flag;
+	int *flag, ret;
 
 	flag = devm_kzalloc(&rpdev->dev, sizeof(int), GFP_KERNEL);
 	*flag = REMOTE_ALLOC;
@@ -796,6 +797,11 @@ static int rtk_rpc_probe(struct rpmsg_device *rpdev)
 	client->big_endian = !rpdev->little_endian;
 
 	client->rpdev = rpdev;
+
+	ret = of_reserved_mem_device_init(client->dev);
+	if (ret)
+		dev_warn(client->dev, "init reserved memory failed");
+
 	mutex_init(&client->r_program_lock);
 	mutex_init(&client->send_lock);
 	spin_lock_init(&client->queue_lock);
-- 
2.47.0

